plutus-core/plutus-ir/src/PlutusIR/Subst.hs

```diff
@@ -1,99 +1,124 @@
-ftvBinding r b = mconcat $ ftvTs ++ ftvTys
- where
-    ftvTs = ftvTerm <$> b^..bindingSubterms
-    ftvTys = ftvTyDataSpecial <$> b^..bindingSubtypes
-
-    -- like ftvTy but specialized for the datatypebind case
-    ftvTyDataSpecial :: Type tyname uni ann -> Set tyname
-    ftvTyDataSpecial ty = case b of
-        DatatypeBind _ (Datatype _ tyconstr tyvars _ _) -> case r of
-            -- for rec, both tyconstr+tyvars are in scope in *WHOLE* dataconstructors
-            Rec -> ftvTy ty \\ setOf bindingTyNames b
-            NonRec ->
-                let tyvarsNames = setOf (traversed.PLC.tyVarDeclName) tyvars
-                    ftvDom = mconcat $ ftvTy <$> funTyArgs ty
-                    -- tyconstr is in scope *only* in the result type codomain
-                    ftvCod = ftvTy (funResultType ty) \\ setOf PLC.tyVarDeclName tyconstr
-                -- for nonrec, the tyvars are in scope in *WHOLE* dataconstructors
-                in (ftvDom <> ftvCod) \\ tyvarsNames
-        _ -> ftvTy ty
+ftvBinding :: Ord tyname => Recursivity -> Traversal' (Binding tyname name uni fun ann) tyname
+ftvBinding r = ftvBindingCtx r mempty
+
+ftvBindingCtx :: Ord tyname => Recursivity -> Set tyname -> Traversal' (Binding tyname name uni fun ann) tyname
+ftvBindingCtx r bound f = \case
+    DatatypeBind a d -> DatatypeBind a <$> ftvDatatypeCtx r bound f d
+    -- sound because the subterms and subtypes are disjoint
+    b                -> ((bindingSubterms . ftvTermCtx bound) `Unsound.adjoin` (bindingSubtypes . ftvTyCtx bound)) f b
+
+ftvDatatypeCtx :: Ord tyname => Recursivity -> Set tyname -> Traversal' (Datatype tyname name uni fun ann) tyname
+ftvDatatypeCtx r bound f d@(Datatype a tyconstr tyvars destr constrs) =
+    let
+        tyConstr = setOf PLC.tyVarDeclName tyconstr
+        tyVars = setOf (traversed.PLC.tyVarDeclName) tyvars
+        allBound = bound `union` tyConstr `union` tyVars
+        varsBound = bound `union` tyVars
+    in case r of
+        -- recursive: introduced names are in scope throughout
+        Rec -> (datatypeSubtypes . ftvTyCtx allBound) f d
+        -- nonrecursive: type constructor is in scope only in the result type of the constructors
+        NonRec ->
+            let
+                combinedTraversal =
+                    -- type arguments are in scope in the argument types
+                    (funArgs . ftvTyCtx varsBound)
+                    -- sound because the argument types and result type are disjoint
+                    `Unsound.adjoin`
+                    -- type constructor and arguments are in scope in the result type
+                    (funRes . ftvTyCtx allBound)
+                constrs' = traverseOf (traversed . PLC.varDeclType . combinedTraversal) f constrs
+            in Datatype a tyconstr tyvars destr <$> constrs'
+
+-- | Traverse the arguments of a function type (nothing if the type is not a function type).
```

Comment:
I think we generally use "function type" to mean just a function type. `forall a . a -> Int` is not a function type, it's a polymorphic type. In Haskell it makes sense to consider these comparable because Haskell has automatic type instantiation, but we don't!